Add one-pager for using machine learning to predict pipeline duration

melotic opened the pull request:
### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/arcade/tree/main/Documentation/Validation

Epic: #8824 

Unfortunately, Github doesnt support inline comments with jupyter notebooks. If you could quote the section of the notebook that you had feedback on for context, that would be much appreciated!

ChadNedzlek commented:
Uh... this doesn't appear to be a one pager... or anything I can read. :-)

ChadNedzlek commented:
Weird. GitHub can _render_ the file more or less fine if you go to "view file".  It's just the diff that's basically worthless. :-)

melotic commented:
> Uh... this doesn't appear to be a one pager... or anything I can read. :-)

Github isn't good with PRs and notebooks. You can click on Files Changed -> three dots on the right -> View File to render it in the browser :/

riarenas commented:
>three dots on the right -> View File to render it in the browser :/

That makes it very unfriendly to reviewers when they want to leave a comment. Let's make a proper one pager doc, and you can link your notebook to it.

ChadNedzlek commented:
This doc is really, really long... Long enough that I'm not sure it should be checked into arcade.  There's like a sentence or two every page that's sometimes saying important stuff... but otherwise I'm not sure what I'm supposed to be getting of this other than "runtime is doing something weird and we should tell them not to do that" and "averages and stddev are stable, so we can just use those".  I'm not sure what the "machine learning" part of this is... it's basically just saying that for a given definition, the build times form a more or less predicable distribution, but that that point, it's just basic statistics.

melotic commented:
> This doc is really, really long... Long enough that I'm not sure it should be checked into arcade. There's like a sentence or two every page that's sometimes saying important stuff... but otherwise I'm not sure what I'm supposed to be getting of this other than "runtime is doing something weird and we should tell them not to do that" and "averages and stddev are stable, so we can just use those". I'm not sure what the "machine learning" part of this is... it's basically just saying that for a given definition, the build times form a more or less predicable distribution, but that that point, it's just basic statistics.

The one-pager outlines the data science process in order to arrive at the conclusion that regressing a distribution over the duration of pipelines is a suitable and effective way to predict the duration of a given pipeline. Without this document, how do we answer the question "How are you giving us prediction?" and "Is your prediction accuarate?" The one-pager (I should rename it something else, and write a seperate markdown one-pager) walks any user through this process and provides suitable statistical analysis and answers to those questions.

While yes, this is basic statistics, the point was that I proved that using basic statistics is effective in solving this problem. The machine learning part comes from the regression on the distrubution.

ChadNedzlek commented:
True... but I'm not sure we need a 10 page doc to answer the question "yes, we did the science, and the science says ok" saved (and fetched by everyone that clones arcade) into our public repository for all time.  A quick summary of the findings in a MD file is likely sufficient.  With a quick summary of "we did analysis of all the repositories, and most of them form a stable normal distribution".

riarenas commented:
Initially Justin came to me with the idea of making his one pager as a jupyter notebook and I thought it was worth the try as long as it was possible to review it. 

We now see that the approach is not correct. This file is too big to check in, and this PR difficult to give feedback on.

@ChadNedzlek's suggestion of just summarizing your findings in the one-pager, along with a link to the full doc for those who want to go through your process sounds right to me.

markwilkie commented:
I love the idea of the jupyter notebook being there because I suspect there's a good amount of learning documented.  However, I guess a giant doc isn't really a "one pager"....lol   So yea, a high level summary of the findings and approach, with the jupyter doc as source material seems great.

melotic commented:
Switched it to markdown.

garath commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
Need a newline between the triple-tick and the word. Otherwise it gets consumed and hidden.

garath commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
_Just_ TimelineBuilds? Does it require any of the companion tables like TimelineRecords?

melotic replied:
Yup. Only TimelineBuilds. The query of how I obtained the data is in the juptyer notebook, but its:

```
TimelineBuilds
| where Result == "succeeded"
| where Project == "public"
| where Reason == "pullRequest"
| where TargetBranch == "main"
| project Repository, Definition, PipelineDuration = datetime_diff('second', FinishTime, StartTime), FinishTime
```

garath commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
How many models? How much data is needed per model?

melotic replied:
I've arbitrarily set the minimum data to be 50 runs for a pipeline to have a model. Each pipeline has its own model.

riarenas replied:
Should this be removed as well since we're no longer using this approach?

garath commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
This implies no cache of already-calculated data. Is that correct?

melotic replied:
If already-calculated you mean the trained models, they are cached in a blob container.

melotic replied:
No longer using Azure Functions.

riarenas replied:
Let's remove it from the doc then

ChadNedzlek commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
Please don't make this an Azure function. Those are a pain to maintain and deploy. Just make this Service Fabric services.

ulisesh replied:
@ChadNedzlek I had the same reaction when Justin mentioned the idea of a new azure function but maintaining Python code somewhere different than an Azure function sounds difficult. I'm curious on what alternatives you have in mind

ChadNedzlek replied:
Pick a model and just code that up in C#/Kusto with a few lines.  I don't think the solution needs to be this complicated, or, at the very least, I don't think the added costs lead to a commensurate benefit.

How _much_ worse are the estimates if we just pick one model and apply it everywhere, for example?

melotic replied:
I'll crunch those numbers. I've already written the timer trigger for fun that builds the models, and its < 100 python LoC, and relatively simple (connect to kusto, download data, train models, save in blob)

I don't envision this being a *huge* project that needs to be maintained, if that's your main concern.

ChadNedzlek replied:
We need to add a python build to the build, we need to add deployment to build, we need to manage a secure NPM flow, and introduce an entirely new package management system, get logging hooked up for python and recorded somewhere, manage secrets. People need to learn python to interact with our project...

It's not the costs of writing the code the first time.  It's all the different infrastructure and other assorted overhead that will need to be created and maintained to make sure the code can live and execute forever.

riarenas replied:
>I don't envision this being a huge project that needs to be maintained, if that's your main concern.

Everything has a maintenance cost. This is something that we need to make sure keeps running, and we need monitoring and troubleshooting in place for it. Chad's considerations are valid in that there has to be good value why we introduce a new piece of infrastructure where we need to keep running, with dependencies up to date, with secrets to keep cycling, etc...

melotic replied:
No need anymore. We'll just implement this with kusto inside queue insights.

ChadNedzlek commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
Why does this need to be so complicated? I must have misunderstood the jypter notebook, because I thought it boiled down to "these things map to a normal distribution, simple STD math works fine", and can just be done with a few lines of Kusto.

melotic replied:
They aren't normally distributed. Some of the follow Beta distributions, other dweibull, some lognorm. We need to pick the best fit for the data.

ChadNedzlek replied:
For which ones does using some STD window not produce reasonable results (other than runtime, which is pretty messed up no matter what model we use)?  Unless doing that math is WILDLY inaccurate, it's far preferable to trying to maintain some python azure function.  The maintenance/security costs of that are pretty high.

melotic replied:
Some examples of when a normal distribution is terrible:

![3715130f-9d3c-4831-bc47-03d0eda490b8](https://user-images.githubusercontent.com/26028834/178365007-cbfbd518-5ae3-4a85-95bf-5ca918c09303.png)
![0e7f97d6-715c-4024-af17-6bd1d7338a1f](https://user-images.githubusercontent.com/26028834/178365019-60a5313b-2b00-48e1-8aec-72e746a43106.png)

We can get much better fits with different PDFs, and thus a better estimate for our customers.

ChadNedzlek replied:
I'm... not sure what that data is saying.  What are we going to tell our customers... and if we just pick a single model and use it everywhere, how does what we're telling the user change?  How often are we "wrong" when we would otherwise have been "right"? By how much are the calculations "off" by?

I understand that mathematically, they are technically better. But what does that end up meaning to the end product? What scenario is impacted, and in what manner, by going with the simpler implementation, so that the time that would be spent delivering the complicated one can be spend delivering value in some other area?

melotic replied:
The normal is a decent fit, even if it looks like it doesnt! We'll just use kusto to compute the normal and stddev and compute the CI.

melotic commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
The trained models are a dictionary, whose keys are the pipeline ID, and values are the fitted probability density function. They'll be pickled and live in a blob container that the Azure Function manages.

melotic replied:
No longer using Azure Functions.

melotic commented:
Testing only fitting normal distributions:
![a3201354-d9c2-4952-97ab-b0aacbd3ec83](https://user-images.githubusercontent.com/26028834/178376299-12b32dae-91ff-4561-8035-21f53d54aa38.png)


Accuracy results:
|       |   Accuracy |
|:------|-----------:|
| count |   29       |
| mean  |   94.9138  |
| std   |    3.14995 |
| min   |   84.7935  |
| 25%   |   93.7422  |
| 50%   |   96.0576  |
| 75%   |   96.9298  |
| max   |   98.3808  |

Statistics on ranges given over backtesting period, in minutes:
|       |         Ranges(min) |
|:------|----------:|
| count | 534       |
| mean  |  26.7273  |
| std   |  24.2594  |
| min   |   4.05823 |
| 25%   |  11.7814  |
| 50%   |  18.9974  |
| 75%   |  26.6553  |
| max   | 107.596   |

@ChadNedzlek seems like this is a good solution.

mathaholic commented:
@melotic you mention overall being concerned about the distributions being different and that being a limitation of your pipeline.  Consider using Chebychev's Inequality Theorem for a model that is more generalizable.  https://statisticsbyjim.com/basics/chebyshevs-theorem-in-statistics/

melotic commented:
> @melotic you mention overall being concerned about the distributions being different and that being a limitation of your pipeline. Consider using Chebychev's Inequality Theorem for a model that is more generalizable. https://statisticsbyjim.com/basics/chebyshevs-theorem-in-statistics/

This solves all of our problems! Thanks Nikki :)

I've updated the one-pager with the new method of doing this. We'll use Chebychev's Inequality and we can do this all in kusto.

melotic commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
We don't show an estimate and instead we just tell them that their pipeline is not predictable.

riarenas commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
Given that the anomaly detection model won't really give much improvement, we should add something here about how we plan to deal with situations such as these. 

For instance, if we know that there's an ongoing infrastructure issue (via the known issues, or some other form of alerting) we could simply decide not to give any predictions.

melotic commented:
Addressed feedback. Would love to get thoughts & feedback on the possible remediations I listed for detecting when Helix or AzDo is on the floor.:

1. Use Grafana's Alert API and check a handful of _bad_ alerts.
2. A test pipeline that periodically runs and sends stuff to Helix that we can monitor.

There's more detail in the one pager (under `Caveats -> Possible Solutions`)

ChadNedzlek commented:
@melotic Why not use the "known issues" stuff that @ulisesh has worked on?  Trying to decide which alerts mean "don't show things" is going to be really difficult, and so rare that it doesn't seem worth coding up.

Honestly, I say just show it, even in weird broken times. Odds are users will have some inclination things are horribly broken... either because of a mail, FR contact, or a "known issue" in the build analysis.  Trying to make this feature smart enough to know that it's not smart enough sounds like a crazy twisted bit of logic that hurts my brain. :-) Just... let it be wrong sometimes and let other mechanisms handle telling uses that "stuff is weird right now".

riarenas commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
Have you considered integrating with the known issues infrastructure in some way? In my mind that's our solution for monitoring issues that could affect repos and global infrastructure.

I see this as less of a "everything is on the floor" detection, as if something is completely down, more things will fail and that will let people know not to expect much from things. But if there is an issue in the repo that breaks it in special ways, it would be good to acknowledge that since there's a known issue, there's not a lot of point in making a prediction

melotic commented:
Known issues is definitely the way to go. I think our best bet is to simply hide the duration if there is a Critical known issue.

riarenas commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
[nit]

```suggestion
There are some issues with this model. First, some pipelines, like `dotnet/runtime`'s have a multimodal distribution. This means we cannot accurately predict their pipeline duration. In this case, their distribution is multimodal because their first step, `Evaluate Paths` evaluates the changes in a given PR, and runs or skips different steps of their pipeline.
```

riarenas commented on Documentation/TeamProcess/One-Pagers/pipeline-machine-learning-arcade8824.md:
I think I would just leave this section out, and instead put the solution to each caveat directly underneath. 

First, some pipeline's like runtime's....

We will handle this by...

riarenas commented:
We should move this doc to live under the other queue insights docs now that we decided to integrate it in the same place. This LGTM with only some minor feedback.
